### PR TITLE
Story/vspc 14

### PR DESCRIPTION
**Pull request by Neehaarika**
Addressed comments and wrote test cases for getModule() method. Please review.

### CODE REVIEW

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_getModule_success() {
+
+        Module newModule = new Module();
+        Optional<Module> findModule = Optional.of(newModule);
+        managerToTest.storeModule(newModule);
+        Mockito.when(moduleRepo.findById(newModule.getId())).thenReturn(findModule);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: getId returns null here, it needs to be set first

Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: also, let's first set up the test then test, meaning first create all the objects you need and tell mockito what to do, then call the method to test, then assert the results. right now you're mixing the order.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        IModule moduleTest = managerToTest.getModule(findModule.get().getId());
+        Assert.assertEquals(moduleTest, newModule);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: actually that's not really what we require, the objects don't have to be the same, but the properties of them have to be. let's test for the id and title.

**Comment by Neehaarika** [comment]
Refactored the test cases for the getModule(String id) based on id as per the method implementation.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java [hunk 3]
```diff
@@ -11,14 +13,27 @@
...
+    @Override
+    public IModule storeModule(IModule module) {
+        return moduleRepo.save((Module) module);
+    }
+
+    @Override
+    public IModule getModule(String id) {
+        boolean isExists = moduleRepo.existsById(id);
+        if (isExists) {
+            return moduleRepo.findById(id).get();
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java]: nope, just do a 'isPresent' on the optional before calling get.

Comment (Neehaarika) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java]: moduleRepo.existById(id) is internally performing a isPresent() check in its method implementation. Also since I faced issue to test that condition in test cases, I changed it to existsById which can be mocked as it is a repo method.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java [hunk 4]
```diff
@@ -11,14 +16,33 @@
...
+
+    @Override
+    public IModule getModule(String id) {
+        Optional<Module> module = moduleRepo.findById(id);
+        if (module.isPresent()) {
+            try {
+                return module.get();
+            } catch(NoSuchElementException e) {
+                logger.error("No Such Element Exception", e);
+            }
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java]: you already checked that there is an object present, so no need to catch the exception

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_getModule_idExists() {
+
+        Module newModule = new Module();
+        Optional<Module> mockModule = Optional.of(newModule);
+        managerToTest.storeModule(newModule);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: why do you call 'storeModule' here?

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void test_getModule_idNotExists() {
+
+        Module newModule = new Module();
+        Optional<Module> mockModule = Optional.of(newModule);
+        Mockito.when(mockModuleRepo.findById("mockString")).thenReturn(mockModule);
+
+        IModule iModuleActual = managerToTest.getModule("mockString");
+        Assert.assertEquals(mockModule.get().getId(), iModuleActual.getId());
+        Assert.assertEquals(mockModule.get().getName(), iModuleActual.getName());
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: you're just testing that the id exists again. but this test is called 'idNotExists'?

**Comment by Neehaarika** [comment]
Please review.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java [hunk 7]
```diff
@@ -11,14 +13,29 @@
...
-	 * @see edu.asu.diging.vspace.core.services.impl.IModuleManager#storeModule(edu.asu.diging.vspace.core.model.IModule, java.lang.String)
-	 */
-	@Override
-	public IModule storeModule(IModule module) {
-		return moduleRepo.save((Module)module);
-	}
+    @Autowired
+    private ModuleRepository moduleRepo;
+    
+    //private final Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java]: remove

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+        IModule iModuleActual = managerToTest.getModule(newModule.getId());
+        Assert.assertEquals(mockModule.get().getId(), iModuleActual.getId());
+        Assert.assertEquals(mockModule.get().getName(), iModuleActual.getName());
+    }
+    
+    @Test(expected = NoSuchElementException.class)
+    public void test_getModule_idNotExists() throws Exception {
+        Mockito.when(mockModuleRepo.findById(Mockito.anyString())).thenReturn(Optional.empty());
+        Mockito.when(mockModuleRepo.findById(Mockito.anyString()).get()).thenThrow(new NoSuchElementException());;
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: when is that thrown? the repository doesn't throw that I believe.

Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: and your manager should just return null if there is no such module (which it does).

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        IModule iModuleActual = managerToTest.getModule(newModule.getId());
+        Assert.assertEquals(mockModule.get().getId(), iModuleActual.getId());
+        Assert.assertEquals(mockModule.get().getName(), iModuleActual.getName());
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void test_getModule_idNotExists() throws Exception {
+        Assert.assertEquals(mockModuleRepo.findById(Mockito.anyString()).get(),
+                managerToTest.getModule(Mockito.anyString()).getId());
+    }
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: this doesn't make sense. You say that a null pointer exception is expected, but the getModule method does not throw a null pointer. You need to test that the method returns null if the id doesn't exist.

**Comment by Neehaarika** [comment]
Please Review.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_getModule_idExists() {
+        Module newModule = new Module();
+        Optional<Module> mockModule = Optional.of(newModule);
+        Mockito.when(mockModuleRepo.findById(newModule.getId())).thenReturn(mockModule);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: newModule.getId() would return null, wouldn't it? findById(null) doesn't make sense, it's not a valid call.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        IModule iModuleActual = managerToTest.getModule(newModule.getId());
+        Assert.assertEquals(mockModule.get().getId(), iModuleActual.getId());
+        Assert.assertEquals(mockModule.get().getName(), iModuleActual.getName());
+    }
+
+    @Test
+    public void test_getModule_idNotExists() throws Exception {
+        Mockito.when(mockModuleRepo.findById(Mockito.anyString())).thenReturn(Optional.empty());
+        Assert.assertEquals(null, managerToTest.getModule(Mockito.anyString()));
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ModuleManagerTest.java]: I believe there there is an assertNull method you can use instead

**Comment by jdamerow** [comment]
test this please
